src/sas/sasgui/perspectives/fitting/gpu_options.py

```diff
@@ -0,0 +1,173 @@
+                if self.sas_opencl == clopt[0]:
+                    button.SetValue(1)
+            else:
+                self.option_button[clopt] = "None"
+                if self.sas_opencl.lower() == "none" :
+                    button.SetValue(1)
+
+            self.Bind(wx.EVT_CHECKBOX, self.on_check, id=button.GetId())
+            self.buttons.append(button)
+            boxsizer.Add(button, 0, 0)
+
+        fit_hsizer = wx.StaticBoxSizer(static_box1, orient=wx.VERTICAL)
+        fit_hsizer.Add(boxsizer, 0, wx.ALL, 5)
+
+        self.panel1.SetSizer(fit_hsizer)
+
+        self.vbox = wx.BoxSizer(wx.VERTICAL)
+        self.vbox.Add(self.panel1, 0, wx.ALL, 10)
+
+        accept_btn = wx.Button(self, wx.ID_OK)
+        accept_btn.SetToolTipString("Accept new OpenCL settings. This will"
+                                    " overwrite SAS_OPENCL variable if set")
+
+        help_id = wx.NewId()
+        help_btn = wx.Button(self, help_id, 'Help')
+        help_btn.SetToolTipString("Help on the GPU options")
+
+        reset_id = wx.NewId()
+        reset_btn = wx.Button(self, reset_id, 'Reset')
+        reset_btn.SetToolTipString("Restore initial settings")
+
+        self.Bind(wx.EVT_BUTTON, self.on_OK, accept_btn)
+        self.Bind(wx.EVT_BUTTON, self.on_reset, reset_btn)
+        self.Bind(wx.EVT_BUTTON, self.on_help, help_btn)
+
+        btn_sizer = wx.BoxSizer(wx.HORIZONTAL)
+        btn_sizer.Add((10, 20), 1) # stretchable whitespace
+        btn_sizer.Add(accept_btn, 0)
+        btn_sizer.Add((10, 20), 0) # non-stretchable whitespace
+        btn_sizer.Add(reset_btn, 0)
+        btn_sizer.Add((10, 20), 0) # non-stretchable whitespace
+        btn_sizer.Add(help_btn, 0)
+
+        # Add the button sizer to the main sizer.
+        self.vbox.Add(btn_sizer, 0, wx.EXPAND|wx.ALL, 10)
+
+        self.SetSizer(self.vbox)
+        self.vbox.Fit(self)
+        self.SetTitle("OpenCL options")
+        self.Centre()
+
+    def _get_clinfo(self):
+        """
+        Reading in information about available OpenCL infrastructure
+        :return:
+        """
+        clinfo = []
+        try:
+            import pyopencl as cl
+            platforms = cl.get_platforms()
+            p_index = 0
+            for platform in platforms:
+                d_index = 0
+                devices = platform.get_devices()
+                for device in devices:
+                    if len(devices) > 1 and len(platforms) > 1:
+                        combined_index = ":".join([str(p_index),str(d_index)])
+                    elif len(platforms) > 1:
+                        combined_index = str(p_index)
+                    else:
+                        combined_index = str(d_index)
+                    #combined_index = ":".join([str(p_index),str(d_index)]) \
+                    #    if len(platforms) > 1 else str(d_index)
+                    clinfo.append((combined_index, ":".join([platform.name,device.name])))
```

Comment:
Please indicate "single precision" if double precision is not available (if "cl_khr_fp64" not in device.extensions).

sasmodels uses the first available double precision device is single = False in the model definition.  If I select the HD 4000 single precision device on my laptop and run a double precision model it falls back to the Intel CPU device, even though I have not selected it in the dialog.

This behaviour should be okay (if the device supports double precision it  can probably run sasview); otherwise we have to make the user select both the single and the double device.  I guess you could list pairs of devices, such as

  X  HD 4000 (single)   Intel CPU (double)
  _   Intel CPU (single)  Intel CPU (double)
  _   No OpenCL

Have you tested this on a card which presents itself as multiple identical devices?